File: pippip-native-ios/More/DeleteAccountCell.m
//
//  DeleteAccountCell.m
//  pippip-native-ios
//
//  Created by Steve Brenneis on 3/21/18.
//  Copyright © 2018 seComm. All rights reserved.
//

#import "DeleteAccountCell.h"
#import "ApplicationSingleton.h"
#import "UserVault.h"
#import "Authenticator.h"
#import <Realm/Realm.h>

@implementation DeleteAccountCell

+ (MoreCellItem*)cellItem {

    MoreCellItem *item = [[MoreCellItem alloc] init];
    item.cellHeight = 45.0;
    item.cellReuseId = @"DeleteAccountCell";
    return item;

}

- (void)awakeFromNib {
    [super awakeFromNib];
    // Initialization code
}

- (void)setSelected:(BOOL)selected animated:(BOOL)animated {
    [super setSelected:selected animated:animated];

    if (selected) {
        [self validateDelete];
    }

}

- (BOOL)deleteUser {

    NSFileManager *manager = [NSFileManager defaultManager];
    RLMRealmConfiguration *config = [RLMRealmConfiguration defaultConfiguration];
    NSArray<NSURL *> *realmFileURLs = @[
                                        config.fileURL,
                                        [config.fileURL URLByAppendingPathExtension:@"lock"],
                                        [config.fileURL URLByAppendingPathExtension:@"note"],
                                        [config.fileURL URLByAppendingPathExtension:@"management"]
                                        ];
    for (NSURL *URL in realmFileURLs) {
        NSError *error = nil;
        [manager removeItemAtURL:URL error:&error];
        if (error) {
            // handle error
        }
    }

    SessionState *sessionState = [ApplicationSingleton instance].accountSession.sessionState;
    NSString *accountName = sessionState.currentAccount;
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
    NSString *docPath = [paths objectAtIndex:0];
    NSString *vaultsPath = [docPath stringByAppendingPathComponent:@"PippipVaults"];
    NSString *vaultPath = [vaultsPath stringByAppendingPathComponent:accountName];
    return [manager removeItemAtPath:vaultPath error:nil];

}

- (void)validateDelete {

    NSString *message = @"You are about to delete this account.\nThis action cannot be undone.\nEnter your passphrase to proceed.";
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"CAUTION!"
                                                                   message:message
                                                            preferredStyle:UIAlertControllerStyleAlert];
    [alert addTextFieldWithConfigurationHandler:^(UITextField* textField) {
        textField.placeholder = @"Passphrase";
        textField.clearButtonMode = UITextFieldViewModeWhileEditing;
    }];
    UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"OK"
                                                       style:UIAlertActionStyleDefault
                                                     handler:^(UIAlertAction *action){
                                                         NSString *passphrase = alert.textFields[0].text;
                                                         [self validatePassphrase:passphrase];
                                                     }];
    UIAlertAction *cancelAction = [UIAlertAction actionWithTitle:@"Cancel"
                                                           style:UIAlertActionStyleCancel
                                                         handler:nil];
    [alert addAction:okAction];
    [alert addAction:cancelAction];
    NSMutableDictionary *info = [NSMutableDictionary dictionary];
    info[@"alert"] = alert;
    [[NSNotificationCenter defaultCenter] postNotificationName:@"PresentAlert" object:nil userInfo:info];

}

- (void)validatePassphrase:(NSString*)passphrase {

    SessionState *sessionState = [ApplicationSingleton instance].accountSession.sessionState;
    NSString *accountName = sessionState.currentAccount;
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
    NSString *docPath = [paths objectAtIndex:0];
    NSString *vaultsPath = [docPath stringByAppendingPathComponent:@"PippipVaults"];
    NSString *vaultPath = [vaultsPath stringByAppendingPathComponent:accountName];
    NSData *vaultData = [NSData dataWithContentsOfFile:vaultPath];
    
    UserVault *vault = [[UserVault alloc] initWithState:sessionState];
    NSError *error = nil;
    [vault decode:vaultData withPassword:passphrase withError:&error];
    if (error == nil) {
        Authenticator *auth = [[Authenticator alloc] init];
        [auth logout];
        [self deleteUser];
        [[NSNotificationCenter defaultCenter] postNotificationName:@"AccountDeleted" object:nil];
    }
    else {
        UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Invalid Passphrase"
                                                                       message:@""
                                                                preferredStyle:UIAlertControllerStyleAlert];
        UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"OK"
                                                           style:UIAlertActionStyleDefault
                                                         handler:nil];
        [alert addAction:okAction];
        NSMutableDictionary *info = [NSMutableDictionary dictionary];
        info[@"alert"] = alert;
        [[NSNotificationCenter defaultCenter] postNotificationName:@"PresentAlert" object:nil userInfo:info];
    }

}

@end

File: pippip-native-ios/More/DeleteAccountCell.swift
//
//  DeleteAccountCell.swift
//  pippip-native-ios
//
//  Created by Steve Brenneis on 4/3/18.
//  Copyright © 2018 seComm. All rights reserved.
//

import UIKit
import PMAlertController
import ChameleonFramework

class DeleteAccountCellItem: MultiCellItemProtocol {

    var cellReuseId: String = "DeleteAccountCell"
    var cellHeight: CGFloat = 45.0
    var currentCell: UITableViewCell?
    
}

class DeleteAccountCell: PippipTableViewCell, MultiCellProtocol {

    static var cellItem: MultiCellItemProtocol = DeleteAccountCellItem()
    var viewController: UITableViewController?
    var accountDeleter = AccountDeleter()
    var sessionState = SessionState()
    var alertPresenter = AlertPresenter()

    override func awakeFromNib() {
        super.awakeFromNib()

        //self.textLabel?.textColor = UIColor.flatRed

    }

    override func setSelected(_ selected: Bool, animated: Bool) {
        super.setSelected(selected, animated: animated)

        if selected {
            validateDelete()
        }

    }

    func validateDelete() {

        let alert = PMAlertController(title: "CAUTION!",
                                      description: "You are about to delete your account\n"
                                                    + "This action cannot be undone\n"
                                                    + "Proceed?",
                                      image: nil,
                                      style: PMAlertControllerStyle.alert)
        alert.addAction(PMAlertAction(title: "Yes",
                                      style: .cancel, action: { () in
                                        DispatchQueue.main.async {
                                            self.checkPassphrase()
                                        }
        }))
        alert.addAction(PMAlertAction(title: "No", style: .default))
        viewController?.present(alert, animated: true, completion: nil)

    }

    func checkPassphrase() {

        let accountName = AccountSession.accountName
        let alert = PMAlertController(title: "Delete Account",
                                      description: "Enter your passphrase",
                                      image: nil,
                                      style: PMAlertControllerStyle.alert)
        alert.addTextField({ (textField) in
            textField?.placeholder = "Passphrase"
            textField?.autocorrectionType = .no
            textField?.spellCheckingType = .no
            textField?.autocapitalizationType = .none
        })
        alert.addAction(PMAlertAction(title: "OK",
                                      style: .default, action: { () in
                                        let passphrase = alert.textFields[0].text ?? ""
                                        do {
                                            if try UserVault.validatePassphrase(passphrase) {
                                                try self.accountDeleter.deleteAccount(accountName: accountName!)
                                                self.alertPresenter.successAlert(title: "Account Deleted",
                                                                                 message: "This account has been deleted and you will now be logged out")
                                                NotificationCenter.default.post(name: Notifications.AccountDeleted,
                                                                                object: nil)
                                            }
                                            else {
                                                self.alertPresenter.infoAlert(title: "Invalid Passphrase",
                                                                              message: "Invalid passphrase, account not deleted")
                                            }
                                        }
                                        catch {
                                            print("Error while deleting account: \(error)")
                                        }
        }))
        alert.addAction(PMAlertAction(title: "Cancel", style: .cancel))
        viewController?.present(alert, animated: true, completion: nil)
        
    }

}
